subscriber_notifee.go

```diff
@@ -0,0 +1,126 @@
+package dht
+
+import (
+	"github.com/jbenet/goprocess"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	"github.com/libp2p/go-eventbus"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// subscriberNotifee implements network.Notifee and also manages the subscriber to the event bus. We consume peer
+// identification events to trigger inclusion in the routing table, and we consume Disconnected events to eject peers
+// from it.
+type subscriberNotifee IpfsDHT
+
+func (nn *subscriberNotifee) DHT() *IpfsDHT {
+	return (*IpfsDHT)(nn)
+}
+
+func (nn *subscriberNotifee) subscribe(proc goprocess.Process) {
+	dht := nn.DHT()
+
+	dht.host.Network().Notify(nn)
+	defer dht.host.Network().StopNotify(nn)
+
+	var err error
+	evts := []interface{}{
+		&event.EvtPeerIdentificationCompleted{},
+	}
+
+	// subscribe to the EvtPeerIdentificationCompleted event which notifies us every time a peer successfully completes identification
+	sub, err := dht.host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+	if err != nil {
+		logger.Errorf("dht not subscribed to peer identification events; things will fail; err: %s", err)
+	}
+	defer sub.Close()
+
+	for {
+		select {
+		case evt, more := <-sub.Out():
+			// we will not be getting any more events
+			if !more {
+				return
+			}
+
+			// something has gone really wrong if we get an event for another type
+			ev, ok := evt.(event.EvtPeerIdentificationCompleted)
```

Comment:
Should also listen for protocol change events? Or are we doing that in a followup?

Reply:
That's in #469 since protocol changes are associated with being able to dynamically switch between client and server modes. Whichever PR gets merged first the other one should make sure to accommodate so that we only use one goroutine for eventbus events instead of two.